app/src/Source.re

```diff
@@ -1,40 +1,68 @@
-module Project = {
-  type project = {
-    address: string,
-    name: string,
-    description: string,
-    imgUrl: string,
-  };
-
-  let mockProjects = [|
-    {
-      address: "monokel",
-      name: "monokel",
-      description: "A looking glass into the future.",
-      imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1557488019/Frame_2_bhz6eq.svg",
-    },
-    {
-      address: "monadic",
-      name: "Monadic",
-      description: "Open source organization of amazing things",
-      imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
-    },
-    {
-      address: "oscoin",
-      name: "open source coin",
-      description: "Infrastructure for the open source community",
-      imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
-    },
-    {
-      address: "radicle",
-      name: "radicle",
-      description: "Decentralized open source collaboration",
-      imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
-    },
-  |];
-
-  let getProjects = (): Js.Promise.t(array(project)) =>
+type address = string;
+
+type project = {
+  address,
+  name: string,
+  description: string,
+  imgUrl: string,
+};
+
+type fetchProjectsResult = Belt.Result.t(array(project), string);
+type registerProjectResult = Belt.Result.t(project, string);
+
+type source = {
+  fetchProjects: unit => Js.Promise.t(fetchProjectsResult),
+  registerProject:
+    (~name: string, ~description: string, ~imgUrl: string) =>
+    Js.Promise.t(registerProjectResult),
+};
+
+let createLocalSource = () => {
+  let mockProjects =
+    ref([|
+      {
+        address: "monokel",
+        name: "monokel",
+        description: "A looking glass into the future.",
+        imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1557488019/Frame_2_bhz6eq.svg",
+      },
+      {
+        address: "monadic",
+        name: "Monadic",
+        description: "Open source organization of amazing things",
+        imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
+      },
+      {
+        address: "oscoin",
+        name: "open source coin",
+        description: "Infrastructure for the open source community",
+        imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
+      },
+      {
+        address: "radicle",
+        name: "radicle",
+        description: "Decentralized open source collaboration",
+        imgUrl: "https://res.cloudinary.com/juliendonck/image/upload/v1549554598/monadic-icon_myhdjk.svg",
+      },
+    |]);
+
+  let fetchProjects = () =>
     Js.Promise.make((~resolve, ~reject as _) =>
-      Js.Global.setTimeout(() => resolve(. mockProjects), 1000) |> ignore
+      Js.Global.setTimeout(
+        () => resolve(. Belt.Result.Ok(mockProjects^)),
```

Comment:
Can you point me to some documentation about the `(. )` and `^` syntax?
Not quite sure what it does here, would love to read up on it.

Reply:
The first `(. )` is to uncurry the function, this [SO answer](https://stackoverflow.com/a/53644093) gives a good overview.

The `^` is to get the actual value held by a ref, more on that [here](https://reasonml.github.io/docs/en/mutation#usage).

Reply:
Woah, that SO answer is deep!